### PR TITLE
override bool(), is_nonzero for real tensor tracing

### PR DESCRIPTION
**Pull request by pianpwk**
Fixes bool() and is_nonzero() calls for real tensor tracing, non-strict export

Differential Revision: D63482693

### CODE REVIEW

**Review by ezyang — CHANGES_REQUESTED** [review]
Did overriding `__bool__` on fake tensor not work?

**Comment by pianpwk** [comment]
> Did overriding `__bool__` on fake tensor not work?

@ezyang Noo, I guess in non-strict we're passing around FunctionalTensors

File: torch/fx/experimental/proxy_tensor.py [hunk 1]
```diff
@@ -801,6 +801,10 @@ def can_handle_tensor(x: Tensor) -> bool:
             if r is not NotImplemented:
                 return r
 
+    if func is torch.ops.aten.is_nonzero.default:
+        with proxy_mode:
+            return (args[0] != 0).item()  # type: ignore[attr-defined]
```
Comment (ezyang) [inline torch/fx/experimental/proxy_tensor.py:806]: This doesn't do anything anymore right?

File: torch/_subclasses/fake_tensor.py [hunk 2]
```diff
@@ -1920,6 +1920,13 @@ def go(t: object, real_t: Tensor) -> None:
                     if isinstance(t.node.expr, sympy.Symbol):
                         assert self.shape_env is not None
                         self.shape_env.set_unbacked_var_to_val(t.node.expr, real_t)
+                    elif (
+                        isinstance(s := t.node.expr, sympy.Eq)
```
Comment (ezyang) [inline torch/_subclasses/fake_tensor.py:1924]: sigh, indicator variables, the gift that keeps giving (no action needed here)

**Review by ezyang — APPROVED** [review]

File: torch/fx/experimental/proxy_tensor.py (hunk 1 above)
Comment (pianpwk) [reply torch/fx/experimental/proxy_tensor.py:806]: I guess this still enables export explicit is_nonzero calls in non-strict + real-tensor mode

**Pull request closed without merging**